asreviewcontrib/insights/algorithms.py

```diff
@@ -21,11 +20,34 @@ def _recall_values(labels, x_absolute=False, y_absolute=False):
 
 
 def _loss_value(labels):
-    positive_doc_ratio = sum(labels) / len(labels)
-    triangle_before_perfect_recall = positive_doc_ratio * 0.5
-    aera_under_recall_curve = metrics.auc(*_recall_values(labels))
-
-    return 1 - (triangle_before_perfect_recall + aera_under_recall_curve)
+    Ny = sum(labels)
+    Nx = len(labels)
+
+    # The best AUC represents the entire area under the perfect curve, which is
+    # the total area Nx * Ny, minus the area above the perfect curve (which is
+    # the sum of a series with a formula (Ny * Ny) / 2) plus 0.5 to account for
+    # the boundary.
+    best_auc = Nx * Ny - (((Ny * Ny) / 2) + 0.5)
```

Comment:
I think this should be `best_auc = Nx * Ny - (Ny * (Ny + 1)) / 2`

Reply:
Hi @fqixiang ,

I agree that this should be the formula, however I am running in to a problem. The values produced aren't in agreement with the calculated AUC.

```python
    def _loss_value(labels):
        Ny = sum(labels)
        Nx = len(labels)

        # The best AUC represents the entire area under the perfect curve, which is
        # the total area Nx * Ny, minus the area above the perfect curve.
        best_auc = Nx * Ny - (((Ny * Ny) / 2) + 0.5)
        new_best_auc = Nx * Ny - (Ny * (Ny + 1)) / 2
        sklearn_best_auc = metrics.auc(*_recall_values(sorted(labels, reverse=True), x_absolute=True, y_absolute=True))

        assert best_auc == sklearn_best_auc, f"{best_auc} is not {sklearn_best_auc} for set {labels}"
>       assert new_best_auc == sklearn_best_auc, f"{new_best_auc} is not {sklearn_best_auc} for set {labels}"
E       AssertionError: 7.0 is not 7.5 for set [1, 1, 0, 0, 0]
```

I am using sklearn.metrics.auc to calculate the area for the best AUC to compare to our implementations. The old formula agrees with sklearn; the new one gives different values. Any ideas?

```python
        best_auc = Nx * Ny - (((Ny * Ny) / 2) + 0.5)
        new_best_auc = (Nx * Ny) - ((Ny * (Ny + 1)) / 2)
        sklearn_best_auc = metrics.auc(*_recall_values(sorted(labels, reverse=True), x_absolute=True, y_absolute=True))

        assert best_auc == sklearn_best_auc, f"{best_auc} is not {sklearn_best_auc} for set {labels}"
>       assert new_best_auc == sklearn_best_auc, f"{new_best_auc} is not {sklearn_best_auc} for set {labels}"
E       AssertionError: 258951.0 is not 258971.5 for set [0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0,...
```

Reply:
Which is ofcourse because the AUC calculation assumes linear instead of stepwise increases. Sometimes just writing it out helps. Thanks, I'm implement the new formulas!

Reply:
And this issue shows up when calculating the actual AUC, since the actual AUC also has this 0.5 over the formula:

```python
def _loss_value(labels):
    Ny = sum(labels)
    Nx = len(labels)

    # The best AUC represents the entire area under the perfect curve, which is
    # the total area Nx * Ny, minus the area above the perfect curve.
    best_auc = Nx * Ny - (Ny * (Ny + 1)) / 2
    print(f"{best_auc = }")

    # Compute recall values (y) based on the provided labels.
    y = np.array(_recall_values(labels, x_absolute=True, y_absolute=True)[1])

    # The actual AUC is calculated by finding the area under the curve using the
    # trapezoidal rule. (y[1:] + y[:-1]) / 2 takes the average height between
    # consecutive y values, and we sum them up.
    actual_auc = np.sum((y[1:] + y[:-1]) / 2)
    print(f"{actual_auc = }")

    # The worst AUC represents the area under the worst-case step curve, which
    # is simply the area under the recall curve where all positive labels are
    # clumped at the end.
    worst_auc = (Ny * Ny) / 2
    print(f"{worst_auc = }")

    # The normalized loss is the difference between the best AUC and the actual
    # AUC, normalized by the range between the best and worst AUCs.
    normalized_loss = (best_auc - actual_auc) / (best_auc - worst_auc) if best_auc != worst_auc else 0  # noqa: E501

    return normalized_loss
```

```
>>> from asreviewcontrib import insights    
>>> insights.metrics._loss_value([1,1,0,0])
best_auc = 5.0
actual_auc = np.float64(5.5)
worst_auc = 2.0
np.float64(-0.16666666666666666)
```

Reply:
Recall curve [0, 0, 1, 2] results in a value of 3 for the worst case. The best case is has recall curve [1, 2, 2, 2], which would be a value of 7 when interpreted in the same manner, but results in a value of 5. I think this is because it actually calculates [0, 1, 2, 2] because of this inversion.

Reply:
And also the actual AUC is just `actual_auc = np.sum(y), no need to average`